### PR TITLE
[JUJU-1083] Make tests upgrade from 2.8

### PR DESCRIPTION
**Pull request by barrettj12**
A re-pull of #14015 which I accidentally closed.

The Upgrade tests should be trying to upgrade from the last minor version - i.e. for 2.9, we should try to upgrade from 2.8.

### CODE REVIEW

**Review by wallyworld — APPROVED** [review]

**Comment by barrettj12** [comment]
It appears Juju 2.8 is having trouble talking to the caas image repo that we set up, I'm not really sure why tho.

File: .github/workflows/smoke.yml [hunk 1]
```diff
@@ -64,12 +64,12 @@ jobs:
     if: github.event.pull_request.draft == false
     strategy:
       matrix:
-        snap_version: ["latest/stable", "latest/beta"]
+        snap_version: ["2.8/stable"]
```
Comment (barrettj12) [inline .github/workflows/smoke.yml:67]: Think we should replace this with an environment variable - don't need the matrix for a single version

File: .github/workflows/smoke.yml [hunk 2]
```diff
@@ -111,8 +112,12 @@ jobs:
 
         echo "::set-output name=go-version::$(grep '^go ' go.mod | awk '{print $2}')"
         echo "::set-output name=base-juju-version::$(juju version | cut -d '-' -f 1)"
-        echo "::set-output name=upstream-juju-version::$(grep -r "const version =" version/version.go | sed -r 's/^const version = \"(.*)\"$/\1/')"
-        echo "::set-output name=juju-db-version::$(grep -r 'DefaultJujuDBSnapChannel =' controller/config.go | sed -r 's/^\s*DefaultJujuDBSnapChannel = \"([[:digit:]]+\.[[:digit:]]+(\.[[:digit:]]+){0,1})\/.*\"$/\1/')"
+        upstreamJujuVersion=$(grep -r "const version =" version/version.go | sed -r 's/^const version = \"(.*)\"$/\1/')
+        echo "::set-output name=upstream-juju-version::${upstreamJujuVersion}"
+        currentStableChannel="$(echo $upstreamJujuVersion | cut -d'.' -f1,2)/stable"
+        currentStableVersion=$(snap info juju | yq ".channels[\"$currentStableChannel\"]" | cut -d' ' -f1)
```
Comment (barrettj12) [inline .github/workflows/smoke.yml:118]: Wait so why do we need `upstreamJujuVersion` **and** `currentStableVersion`?

File: .github/workflows/smoke.yml [hunk 3]
```diff
@@ -212,15 +218,28 @@ jobs:
 
         sg microk8s <<EOF
           juju bootstrap microk8s c \
-            --config caas-image-repo='{"repository": "${DOCKER_REGISTRY}/test-repo", "serveraddress": "https://${DOCKER_REGISTRY}/"}' \
-            --config juju-db-snap-channel="${JUJU_DB_TAG}/stable" \
-            --config features="[developer-mode]"
+            --config caas-image-repo="${DOCKER_REGISTRY}/test-repo" \
+            --config features="[developer-mode]" --debug --keep-broken
```
Comment (barrettj12) [inline .github/workflows/smoke.yml:222]: Not sure if we should keep the --debug and --keep-broken flags

File: .github/workflows/smoke.yml [hunk 4]
```diff
@@ -283,11 +313,29 @@ jobs:
...
+          fi
+          sleep 10
+          attempt=$((attempt+1))
+          if [ "$attempt" -eq 48 ]; then
+              echo "Upgrade controller timed out"
+              exit 1
+          fi
+        done
+        
         juju upgrade-controller ${UPGRADE_PARAMS_${{ matrix.model_type }}}
```
Comment (barrettj12) [inline .github/workflows/smoke.yml:339]: This part should be removed now right?

File: apiserver/facades/client/client/client.go [hunk 5]
```diff
@@ -881,7 +881,7 @@ func (c *Client) toolVersionsForCAAS(args params.FindToolsParams, streamsVersion
 			}
 		}
 		arch, err := reg.GetArchitecture(imageName, number.String())
-		if errors.IsNotFound(err) {
+		if errors.IsNotFound(errors.Cause(err)) {
```
Comment (barrettj12) [inline apiserver/facades/client/client/client.go:884]: What's the rationale for this - and maybe it should go in a separate PR?

File: .github/workflows/smoke.yml (hunk 1 above)
Comment (ycliuhw) [reply .github/workflows/smoke.yml:67]: I think it's better to keep it as an array and we should probably consider to add `["2.8/stable", "2.9/stable"]` for develop branch.

File: .github/workflows/smoke.yml (hunk 2 above)
Comment (ycliuhw) [reply .github/workflows/smoke.yml:118]: Because we have to upgrade from currentStableVersion then upstreamJujuVersion.
You see the two upgrade steps?

File: .github/workflows/smoke.yml (hunk 4 above)
Comment (ycliuhw) [reply .github/workflows/smoke.yml:339]: Nah, we have to upgrade to stable first then local version.

File: apiserver/facades/client/client/client.go (hunk 5 above)
Comment (ycliuhw) [reply apiserver/facades/client/client/client.go:884]: Yeah, this was for testing purposes. It will be in a separate PR with tests,

File: .github/workflows/smoke.yml [hunk 6]
```diff
@@ -125,7 +130,7 @@ jobs:
       if: env.RUN_TEST == 'RUN' && matrix.model_type == 'microk8s'
       uses: balchua/microk8s-actions@v0.2.2
       with:
-        channel: "latest/stable"
+        channel: "1.23/stable"
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply .github/workflows/smoke.yml:133]: Changed to `latest/stable`
